### PR TITLE
Directional rounding

### PR DESCRIPTION
**Pull request by Rubilmax**
This is a proposal for #55. Please bear in mind that:
- modifying existing library functions is a breaking change for our open-source library
- expecting the dev to append a rounding direction to wad/ray operations is cumbersome

For this reason, I suggest to append `wadMulDown` & `wadMulUp` implementations instead

I'm expecting you to tell me if it's the correct way to go or not, so that I continue working on this, following this checklist:
- [x] division operation
- [x] ray scale
- [x] percentage scale
- [x] formal proof using Certora

### CODE REVIEW

File: src/math/WadRayMath.sol [hunk 1]
```diff
@@ -10,12 +10,14 @@ library WadRayMath {
...
     uint256 internal constant WAD = 1e18;
     uint256 internal constant HALF_WAD = 0.5e18;
+    uint256 internal constant WAD_MINUS_ONE = 1e18 - 1;
     uint256 internal constant RAY = 1e27;
     uint256 internal constant HALF_RAY = 0.5e27;
     uint256 internal constant WAD_RAY_RATIO = 1e9;
     uint256 internal constant HALF_WAD_RAY_RATIO = 0.5e9;
     uint256 internal constant MAX_UINT256 = 2**256 - 1; // Not possible to use type(uint256).max in yul.
     uint256 internal constant MAX_UINT256_MINUS_HALF_WAD = 2**256 - 1 - 0.5e18;
+    uint256 internal constant MAX_UINT256_MINUS_WAD_MINUS_ONE = 2**256 - 1 - (1e18 - 1);
```
Comment (MerlinEgalite) [inline src/math/WadRayMath.sol:20]:
```suggestion
    uint256 internal constant MAX_UINT256_MINUS_WAD_MINUS_ONE = 2**256 - WAD;
```

**Review by MerlinEgalite — APPROVED** [review]

File: src/math/WadRayMath.sol (hunk 1 above)
Comment (Rubilmax) [reply src/math/WadRayMath.sol:20]: You can't use composed constants in Yul. I'm updating the comment above to make it clear

Comment (MerlinEgalite) [reply src/math/WadRayMath.sol:21]: Ok then just `2**256 - 1e18`?

Comment (Rubilmax) [reply src/math/WadRayMath.sol:21]: That would work too and I'm ok with it, though it's less obvious to the reader

Comment (MerlinEgalite) [reply src/math/WadRayMath.sol:21]: Not a blocker we can leave as it is np

File: src/math/WadRayMath.sol [hunk 2]
```diff
@@ -8,19 +8,26 @@ pragma solidity ^0.8.0;
...
+    uint256 internal constant RAY_MINUS_ONE = 1e27 - 1;
+    uint256 internal constant RAY_WAD_RATIO = 1e9;
+    uint256 internal constant HALF_RAY_WAD_RATIO = 0.5e9;
+    uint256 internal constant MAX_UINT256 = 2**256 - 1;
     uint256 internal constant MAX_UINT256_MINUS_HALF_WAD = 2**256 - 1 - 0.5e18;
     uint256 internal constant MAX_UINT256_MINUS_HALF_RAY = 2**256 - 1 - 0.5e27;
+    uint256 internal constant MAX_UINT256_MINUS_WAD_MINUS_ONE = 2**256 - 1 - (1e18 - 1);
+    uint256 internal constant MAX_UINT256_MINUS_RAY_MINUS_ONE = 2**256 - 1 - (1e27 - 1);
+    uint256 internal constant MAX_UINT256_WAD_RATIO = 0x12725dd1d243aba0e75fe645cc4873f9e65afe688c928e1f21;
+    uint256 internal constant MAX_UINT256_RAY_RATIO = 0x4f3a68dbc8f03f243baf513267aa9a3ee524F8e028;
```
Comment (Rubilmax) [inline src/math/WadRayMath.sol:26]: we can't write `(2**256 - 1) / 1e18` nor compose constants here

**Review by pakim249CAL — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Review by pakim249CAL — APPROVED** [review]

**Comment by MathisGD** [comment]
Please let me the time to review it tomorrow

File: src/math/WadRayMath.sol [hunk 3]
```diff
@@ -57,7 +99,43 @@ library WadRayMath {
...
+    /// @dev Divides two wad, flooring the result.
+    /// @param x Wad.
+    /// @param y Wad.
+    /// @return z The result of x / y, in wad.
+    function wadDivDown(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        // Overflow if y == 0
+        // Overflow if (x * WAD) > type(uint256).max
+        // <=> x > type(uint256).max / WAD
+        assembly {
+            if iszero(mul(y, iszero(gt(x, MAX_UINT256_WAD_RATIO)))) {
```
Comment (MathisGD) [inline src/math/WadRayMath.sol:111]:
```suggestion
            if iszero(mul(y, lt(x, MAX_UINT256_WAD_RATIO_PLUS_ONE))) {
```

Comment (MathisGD) [inline src/math/WadRayMath.sol:111]:
```suggestion
            if iszero(mul(y, lt(x, MAX_UINT256_WAD_RATIO_PLUS_ONE))) {
```

File: src/math/WadRayMath.sol [hunk 4]
```diff
@@ -94,13 +207,49 @@ library WadRayMath {
...
+    /// @dev Divides two ray, flooring the result.
+    /// @param x Wad.
+    /// @param y Wad.
+    /// @return z The result of x / y, in ray.
+    function rayDivDown(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        // Overflow if y == 0
+        // Overflow if (x * RAY) > type(uint256).max
+        // <=> x > type(uint256).max / RAY
+        assembly {
+            if iszero(mul(y, iszero(gt(x, MAX_UINT256_RAY_RATIO)))) {
```
Comment (MathisGD) [inline src/math/WadRayMath.sol:219]:
```suggestion
            if iszero(mul(y, lt(x, MAX_UINT256_RAY_RATIO_PLUS_ONE))) {
```

**Review by MathisGD — COMMENTED** [review]
I would also harmonize the comments about rounding for already existing functions (#58)

File: src/math/WadRayMath.sol (hunk 2 above)
Comment (Rubilmax) [reply src/math/WadRayMath.sol]: ahah yes

File: src/math/PercentageMath.sol [hunk 5]
```diff
@@ -99,6 +138,42 @@ library PercentageMath {
...
+    /// @notice Executes a percentage division (x / p), flooring the result.
+    /// @param x The value to divide by the percentage.
+    /// @param percentage The percentage of the value to divide.
+    /// @return y The result of the division.
+    function percentDivDown(uint256 x, uint256 percentage) internal pure returns (uint256 y) {
+        // Overflow if percentage == 0
+        // Overflow if (x * PERCENTAGE_FACTOR) > type(uint256).max
+        // <=> x > type(uint256).max / PERCENTAGE_FACTOR
+        assembly {
+            if iszero(mul(percentage, iszero(gt(x, MAX_UINT256_PERCENTAGE_FACTOR_RATIO)))) {
```
[the remark this answers is not in the document]
Comment (Rubilmax) [reply src/math/PercentageMath.sol]: opened an issue: https://github.com/morpho-dao/morpho-utils/issues/85

**Review by MerlinEgalite — APPROVED** [review]

File: src/math/PercentageMath.sol [hunk 6]
```diff
@@ -78,6 +82,41 @@ library PercentageMath {
         }
     }
 
+    /// @notice Executes a percentage multiplication (x * p), flooring the result.
```
[the remark this answers is not in the document]
Comment (MathisGD) [reply src/math/PercentageMath.sol]: You don't like it ?

File: src/math/PercentageMath.sol [hunk 7]
```diff
@@ -99,6 +138,42 @@ library PercentageMath {
         }
     }
 
+    /// @notice Executes a percentage division (x / p), flooring the result.
+    /// @param x The value to divide by the percentage.
+    /// @param percentage The percentage of the value to divide.
+    /// @return y The result of the division.
+    function percentDivDown(uint256 x, uint256 percentage) internal pure returns (uint256 y) {
+        // Overflow if percentage == 0
```
[the remark this answers is not in the document]
Comment (MathisGD) [reply src/math/PercentageMath.sol]: I think that this should be fixed

File: src/math/PercentageMath.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (Rubilmax) [reply src/math/PercentageMath.sol]: Yes I do and it is applied!

[the remark this answers is not in the document]
Comment (MathisGD) [reply src/math/PercentageMath.sol]: do I have a bug or what ?

Capture d’écran 2023-01-02 à 12 12 57

[the remark this answers is not in the document]
Comment (Rubilmax) [reply src/math/PercentageMath.sol]: If you click on the title of the conversation, it redirects you to the commit at which you opened the conversation, which is not the latest (and thus the fix is not yet applied)
If you wanna see the fix, you need to open the changes tab
Also, there's a little `outdated` label next to the conversation's title indicating the code was changed
image

**Review by MathisGD — APPROVED** [review]

File: src/math/PercentageMath.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (MathisGD) [reply src/math/PercentageMath.sol]: ofc.. sorry

**Pull request merged**